src/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp

```diff
@@ -66,48 +69,67 @@ APPLICATION_INFO::CreateHandler(
 HRESULT
 APPLICATION_INFO::CreateApplication(const IHttpApplication& pHttpApplication)
 {
-    HRESULT hr = S_OK;
-
     if (AppOfflineApplication::ShouldBeStarted(pHttpApplication))
     {
         LOG_INFO(L"Detected app_offline file, creating polling application");
-        #pragma warning( push )
-        #pragma warning ( disable : 26409 ) // Disable "Avoid using new", using custom deleter here
-        m_pApplication.reset(new AppOfflineApplication(pHttpApplication));
-        #pragma warning( pop )
+        m_pApplication.reset(std::make_unique<AppOfflineApplication>(pHttpApplication).release());
+        return S_OK;
     }
     else
     {
-        FINISHED_IF_FAILED(m_handlerResolver.GetApplicationFactory(pHttpApplication, m_pApplicationFactory));
+        try
+        {
+            const WebConfigConfigurationSource configurationSource(m_pServer.GetAdminManager(), pHttpApplication);
+            ShimOptions options(configurationSource);
 
-        LOG_INFO(L"Creating handler application");
-        IAPPLICATION * newApplication;
-        FINISHED_IF_FAILED(m_pApplicationFactory->Execute(
-            &m_pServer,
-            &pHttpApplication,
-            &newApplication));
+            const auto hr = TryCreateApplication(pHttpApplication, options);
 
-        m_pApplication.reset(newApplication);
+            if (FAILED_LOG(hr))
+            {
+                // Log the failure and update application info to not try again
+                EventLog::Error(
+                    ASPNETCORE_EVENT_ADD_APPLICATION_ERROR,
+                    ASPNETCORE_EVENT_ADD_APPLICATION_ERROR_MSG,
+                    pHttpApplication.GetApplicationId(),
+                    hr);
+
+                m_pApplication.reset(std::make_unique<ServerErrorApplication>(
+                    pHttpApplication,
+                    hr,
+                    options.QueryDisableStartupPage(),
+                    options.QueryHostingModel() == APP_HOSTING_MODEL::HOSTING_IN_PROCESS ? IN_PROCESS_SHIM_STATIC_HTML : OUT_OF_PROCESS_SHIM_STATIC_HTML)
+                    .release());
+            }
+        }
+        catch (ConfigurationLoadException &ex)
+        {
+            EventLog::Error(
+                ASPNETCORE_CONFIGURATION_LOAD_ERROR,
+                ASPNETCORE_CONFIGURATION_LOAD_ERROR_MSG,
+                ex.get_message().c_str());
+        
+            m_pApplication.reset(std::make_unique<ServerErrorApplication>(pHttpApplication, E_FAIL /* hr */).release());
```

Comment:
Do we need a third error page for config load/other generic failures?

Reply:
We don't know whether we are allowed to return startup pages as it is an option in config.

Reply:
Ah, right